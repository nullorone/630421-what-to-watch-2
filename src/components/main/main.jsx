import React from "react";
import PropTypes from 'prop-types';
import MovieCardSmallList from "../movie-card-small-list/movie-card-small-list";
import GenreList from "../genre-list/genre-list";
import IconsWrapper from "../icons-wrapper/icons-wrapper";
import Logo from "../logo/logo";
import MovieCard from "../movie-card/movie-card";

const {string, shape, number, bool, arrayOf, func} = PropTypes;

const Main = (props) => {
  const {
    films,
    promo,
    genres,
    icons,
    selectedGenre,
    onSelectedGenreClick,
  } = props;

  return (
    <>
      {icons && <IconsWrapper iconNames={icons}/>}

      <MovieCard {...promo}/>

      <div className="page-content">
        <section className="catalog">
          <h2 className="catalog__title visually-hidden">
            Catalog
          </h2>

          {genres
          && <GenreList
            genres={genres}
            selectedGenre={selectedGenre}
            onGenreClick={onSelectedGenreClick}/>}

          {films && <MovieCardSmallList films={films}/>}
        </section>

        <footer className="page-footer">
          <Logo light={true}/>
          <div className="copyright">
            <p>Â© 2019 What to watch Ltd.</p>
          </div>
        </footer>
      </div>
    </>
  );
};

Main.propTypes = {
  films: arrayOf(shape({
    id: number.isRequired,
    name: string.isRequired,
    image: shape({
      poster: string.isRequired,
      posterAlt: string.isRequired,
      preview: string.isRequired,
      previewAlt: string.isRequired,
      background: string.isRequired,
    }),
    backgroundColor: string.isRequired,
    video: shape({
      link: shape({
        mp4: string.isRequired,
        webm: string.isRequired,
      }),
      poster: string.isRequired,
    }),
    description: string.isRequired,
    rating: number.isRequired,
    scoresCount: number.isRequired,
    director: string.isRequired,
    starring: arrayOf(string.isRequired),
    runTime: number.isRequired,
    genre: string.isRequired,
    released: number.isRequired,
    isFavorite: bool.isRequired,
  })),
  promo: shape({
    id: number.isRequired,
    name: string.isRequired,
    image: shape({
      poster: string.isRequired,
      posterAlt: string.isRequired,
      preview: string.isRequired,
      previewAlt: string.isRequired,
      background: string.isRequired,
    }),
    backgroundColor: string.isRequired,
    video: shape({
      link: string.isRequired,
      preview: string.isRequired,
    }),
    description: string.isRequired,
    rating: number.isRequired,
    scoresCount: number.isRequired,
    director: string.isRequired,
    starring: arrayOf(string.isRequired),
    runTime: number.isRequired,
    genre: string.isRequired,
    released: number.isRequired,
    isFavorite: bool.isRequired,
  }),
  genres: arrayOf(string.isRequired),
  selectedGenre: string.isRequired,
  onSelectedGenreClick: func.isRequired,
  icons: arrayOf(string.isRequired),
};

export default Main;
