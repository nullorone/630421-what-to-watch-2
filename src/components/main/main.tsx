import * as React from "react";
import GenreList from "../genre-list/genre-list";
import IconsWrapper from "../icons-wrapper/icons-wrapper";
import Logo from "../logo/logo";
import MovieCard from "../movie-card/movie-card";
import {Film, UserData} from "../../types";
import {iconNames} from "../../constants";
import withAddItemButton from "../../hocs/with-add-item-button/with-add-item-button";
import withChangeItem from "../../hocs/with-change-item/with-change-item";
import MovieCardSmallList from "../movie-card-small-list/movie-card-small-list";

interface MainProps {
  promo: Film;
  filteredFilms: Film[];
  films: Film[];
  genres: string[];
  selectedGenre: string;
  onSelectedGenreClick: () => void;
  onListClick?: (favorite: boolean) => void;
  onButtonClick: () => void;
  user: UserData;
}

const Main: React.FC<MainProps> = (props) => {
  const {
    promo,
    genres,
    selectedGenre,
    onSelectedGenreClick,
    onButtonClick,
    onListClick,
    user,
    filteredFilms,
    films,
  } = props;

  const MovieCardSmallListWrapped = withAddItemButton(withChangeItem(MovieCardSmallList));

  return (
    <>
      <IconsWrapper iconNames={iconNames}/>

      <MovieCard {...promo} user={user} onPlayButtonClick={onButtonClick} onListClick={onListClick}/>

      <div className="page-content">
        <section className="catalog">
          <h2 className="catalog__title visually-hidden">
            Catalog
          </h2>

          {genres
          && <GenreList
            genres={genres}
            selectedGenre={selectedGenre}
            onGenreClick={onSelectedGenreClick}/>}

          <MovieCardSmallListWrapped
            films={(filteredFilms.length !== 0)
              ? filteredFilms
              : films
            }/>
        </section>

        <footer className="page-footer">
          <Logo light={true}/>
          <div className="copyright">
            <p>Â© 2019 What to watch Ltd.</p>
          </div>
        </footer>
      </div>
    </>
  );
};

export default Main;
