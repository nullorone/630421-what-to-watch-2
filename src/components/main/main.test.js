import React from 'react';
import renderer from "react-test-renderer";
import {Main} from './main';

const MOCK_MOVIES = [{
  title: ``,
  img: {
    src: ``,
    alt: ``,
  },
}];

it(`Render main screen`, () => {
  const onMovieCardLinkClick = jest.fn();
  const initialProps = {
    movies: MOCK_MOVIES,
    onMovieCardLinkClick,
  };

  const tree = renderer
    .create(
        <Main
          {...initialProps}
        />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
