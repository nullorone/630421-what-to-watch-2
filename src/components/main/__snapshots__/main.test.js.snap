// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Make snapshot Render main screen 1`] = `
Array [
  <IconsWrapper
    iconNames={
      Array [
        "PAUSE",
      ]
    }
  />,
  <MovieCard
    backgroundColor=""
    description=""
    director="Keks"
    genre=""
    id={0}
    image={
      Object {
        "background": "",
        "backgroundAlt": "",
        "poster": "",
        "posterAlt": "1",
        "preview": "",
        "previewAlt": "",
      }
    }
    isFavorite={true}
    name="Keks"
    rating={1}
    released={1}
    runTime={0}
    scoresCount={1}
    starring={
      Array [
        "Me",
        "You",
        "They",
      ]
    }
    video={
      Object {
        "link": "",
        "preview": "",
      }
    }
  />,
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <GenreList
        genres={
          Array [
            "ECMA",
            "CSS",
          ]
        }
        onGenreClick={[MockFunction]}
        selectedGenre=""
      />
      <MovieCardSmallList
        films={
          Array [
            Object {
              "backgroundColor": "",
              "description": "",
              "director": "Keks",
              "genre": "",
              "id": 0,
              "image": Object {
                "background": "",
                "backgroundAlt": "",
                "poster": "",
                "posterAlt": "1",
                "preview": "",
                "previewAlt": "",
              },
              "isFavorite": true,
              "name": "Keks",
              "rating": 1,
              "released": 1,
              "runTime": 0,
              "scoresCount": 1,
              "starring": Array [
                "Me",
                "You",
                "They",
              ],
              "video": Object {
                "link": Object {
                  "mp4": "",
                  "webm": "",
                },
                "poster": "",
              },
            },
          ]
        }
      />
    </section>
    <footer
      className="page-footer"
    >
      <Logo
        light={true}
      />
      <div
        className="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
